subsys/fs/ext2/ext2_disk_access.c

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (c) 2023 Antmicro <www.antmicro.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log.h>
+#include <zephyr/device.h>
+#include <zephyr/storage/disk_access.h>
+
+#include "ext2.h"
+#include "ext2_struct.h"
+
+LOG_MODULE_DECLARE(ext2);
+
+static struct disk_data {
+	const char *name;
+	uint32_t sector_size;
+	uint32_t sector_count;
+} disk_data;
+
+static int64_t disk_access_device_size(struct ext2_data *fs)
+{
+	struct disk_data *disk = fs->backend;
+
+	return disk->sector_count * disk->sector_size;
+}
+
+static int64_t disk_access_write_size(struct ext2_data *fs)
+{
+	struct disk_data *disk = fs->backend;
+
+	return disk->sector_size;
+}
+
+static int disk_read(const char *disk, uint8_t *buf, uint32_t start, uint32_t num)
+{
+	int rc, loop = 0;
+
+	do {
+		rc = disk_access_ioctl(disk, DISK_IOCTL_CTRL_SYNC, NULL);
+		if (rc == 0) {
+			rc = disk_access_read(disk, buf, start, num);
+			LOG_DBG("disk read: (start:%d, num:%d) (ret: %d)", start, num, rc);
+		}
+	} while ((rc == -EBUSY) && (loop++ < 16));
+	return rc;
+}
+
+static int disk_write(const char *disk, const uint8_t *buf, uint32_t start, uint32_t num)
+{
+	int rc, loop = 0;
+
+	do {
+		rc = disk_access_ioctl(disk, DISK_IOCTL_CTRL_SYNC, NULL);
+		if (rc == 0) {
+			rc = disk_access_write(disk, buf, start, num);
+			LOG_DBG("disk write: (start:%d, num:%d) (ret: %d)", start, num, rc);
+		}
+	} while ((rc == -EBUSY) && (loop++ < 16));
+	return rc;
+}
+
+static int disk_prepare_range(struct disk_data *disk, uint32_t addr, uint32_t size,
+		uint32_t *s_start, uint32_t *s_count)
+{
+	*s_start = CONFIG_EXT2_DISK_STARTING_SECTOR + addr / disk->sector_size;
+	*s_count = size / disk->sector_size;
+
+	LOG_DBG("addr:0x%x size:0x%x -> sector_start:%d sector_count:%d",
+			addr, size, *s_start, *s_count);
+
+	/* Cannot read or write outside the disk. */
+	if (*s_start + *s_count > disk->sector_count) {
+		LOG_ERR("Requested sectors: %d-%d are outside of disk (num_sectors: %d)",
+			*s_start, *s_start + *s_count, disk->sector_count);
+		return -ENOSPC;
+	}
```

Comment:
Question: Is it worth checking for  a 32-bit overflow? This could happen if CONFIG_EXT2_DISK_STARTING_SECTOR is sufficiently large.

Reply:
I've added a check for overflow. I thought that it may be better to just change these values to 64-bits values, but `disk_access` does support 32-bit integers, so check is now the only thing we can do.